go-controller/pkg/metrics/master.go

```diff
@@ -693,6 +728,520 @@ func getPodUIDFromPortBinding(row *sbdb.PortBinding) kapimtypes.UID {
+			nbCfg: nbGlobal.NbCfg + 1})
+		cr.measurements[kindNamespaceName] = m
+		cr.measurementsMu.Unlock()
+	}, ovnStartTimestamp, nil
+}
+
+// runMeasurementRateAdjuster will adjust the rate of measurements based on the number of nodes in the cluster and arg k
+func (cr *ConfigDurationRecorder) runMeasurementRateAdjuster(kube kube.Interface, k float64, nodeCheckPeriod time.Duration,
+	stop <-chan struct{}) {
+	var currentMeasurementRate, newMeasurementRate uint64
+
+	updateMeasurementRate := func() {
+		if nodeCount, err := getNodeCount(kube); err != nil {
+			klog.Errorf("Config duration recorder: failed to update ticker duration considering node count: %v", err)
+		} else {
+			newMeasurementRate = uint64(math.Round(k * float64(nodeCount)))
+			if newMeasurementRate != currentMeasurementRate {
+				if newMeasurementRate > 0 {
+					currentMeasurementRate = newMeasurementRate
+					cr.measurementRate = newMeasurementRate
+				}
+				klog.V(5).Infof("Config duration recorder: updated measurement rate to approx 1 in"+
+					" every %d requests", newMeasurementRate)
+			}
+		}
+	}
+
+	// initial measurement rate adjustment
+	updateMeasurementRate()
+
+	go func() {
+		nodeCheckTicker := time.NewTicker(nodeCheckPeriod)
+		for {
+			select {
+			case <-nodeCheckTicker.C:
+				updateMeasurementRate()
+			case <-stop:
+				nodeCheckTicker.Stop()
+				return
+			}
+		}
+	}()
+}
+
+// processMeasurements manages the measurements map. It calculates metrics and cleans up finished or stale measurements
+func (cr *ConfigDurationRecorder) processMeasurements(period time.Duration, updateOVNMeasurementCh chan hvCfgUpdate,
+	deleteOVNMeasurementCh chan int, stop <-chan struct{}) {
+	ticker := time.NewTicker(period)
+	var ovnKDelta, ovnDelta float64
+
+	for {
+		select {
+		case <-stop:
+			ticker.Stop()
+			return
+		// remove measurements if channel updateOVNMeasurementCh overflows, therefore we cannot trust existing measurements
+		case hvCfg := <-deleteOVNMeasurementCh:
+			cr.measurementsMu.Lock()
+			removeOVNMeasurements(cr.measurements, hvCfg)
+			cr.measurementsMu.Unlock()
+		case h := <-updateOVNMeasurementCh:
+			cr.measurementsMu.Lock()
+			cr.addHvCfg(h.hvCfg, h.timestamp)
+			cr.measurementsMu.Unlock()
+		// used for processing measurements that didn't require OVN measurement. Helps to keep measurement map small
+		case kindNamespaceName := <-cr.triggerProcessCh:
+			cr.measurementsMu.Lock()
+			m, ok := cr.measurements[kindNamespaceName]
+			if !ok {
+				klog.Errorf("Config duration recorder: expected measurement, but not found")
+				cr.measurementsMu.Unlock()
+				continue
+			}
+			if !m.end {
+				cr.measurementsMu.Unlock()
+				continue
+			}
+			if len(m.ovnMeasurements) != 0 {
+				cr.measurementsMu.Unlock()
+				continue
+			}
+			ovnKDelta = m.endTimestamp.Sub(m.startTimestamp).Seconds()
+			metricNetworkProgramming.With(prometheus.Labels{"kind": m.kind}).Observe(ovnKDelta)
```

Comment:
metricNetworkProgramming then can include ovn-k only measurements, combined with e2e measurements, why would we do that?
Also, why have you decied to switch from only calling AddOVN with startTimestamp to more complicated Start->AddOVN->End chain of calls?

Reply:
>metricNetworkProgramming then can include ovn-k only measurements, combined with e2e measurements, why would we do that?

I want to capture the full e2e duration for each kind. This metric has lower value if I separate it out into just the CMS and OVN as folks care about the e2e duration more than just CMS or OVN alone.
If folks want to look at the OVN duration, there is a dedicated metric for that independent of Kind.

>Also, why have you decied to switch from only calling AddOVN with startTimestamp to more complicated Start->AddOVN->End chain of calls?

I didn't like passing the timestamp through numerous layers of functions. Also, I couldnt start measuring before workqueues.